scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/core/SchedulerFrontendState.java

```diff
@@ -598,27 +615,32 @@ synchronized void disconnect() throws NotConnectedException, PermissionException
      * @param id
      *            the uniqueID of the user
      */
-    private synchronized void disconnect(UniqueID id) {
-        credentials.remove(id);
-        ListeningUser ident = identifications.remove(id);
-        if (ident != null) {
-            // remove listeners if needed
-            ident.clearListener();
-            // remove this user to the list of connected user if it has not
-            // already been removed
-            ident.getUser().setToRemove();
-            sState.getUsers().update(ident.getUser());
-            // cancel the timer
-            ident.getUser().getSession().cancel();
-            // log and send events
-            String user = ident.getUser().getUsername();
-            logger.info("User '" + user + "' has disconnect the scheduler !");
-            dispatchUsersUpdated(new NotificationData<UserIdentification>(SchedulerEvent.USERS_UPDATE, ident.getUser()),
-                                 false);
-        }
+    private void disconnect(UniqueID id) {
+        Lambda.withLockInterruptible(usersWriteLock, () -> {
+            credentials.remove(id);
+            ListeningUser ident = identifications.remove(id);
+            if (ident != null) {
+                // remove listeners if needed
+                ident.clearListener();
+                // remove this user to the list of connected user if it has not
+                // already been removed
+                ident.getUser().setToRemove();
+                Lambda.withLock(stateWriteLock, () -> sState.getUsers().update(ident.getUser()));
```

Comment:
We could add a comment in the declaration of userLock and stateLock about their order, to reduce risk of deadlock in the future dev.

Reply:
Very good remark, and indeed I made a mistake somewhere regarding lock hierarchy